src/python/pants/core/goals/publish.py

```diff
@@ -0,0 +1,186 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from pants.engine.unions import UnionRule
+from typing_extensions import final
+from typing import TypeVar, ClassVar, Generic, Type
+import logging
+from abc import ABCMeta
+from dataclasses import dataclass, field, replace
+from itertools import chain
+
+from pants.core.goals.package import BuiltPackage, PackageFieldSet
+from pants.engine.console import Console
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.process import InteractiveProcess, InteractiveRunner
+from pants.engine.rules import Get, MultiGet, collect_rules, goal_rule, rule, Rule
+from pants.engine.target import (
+    FieldSet,
+    NoApplicableTargetsBehavior,
+    TargetRootsToFieldSets,
+    TargetRootsToFieldSetsRequest,
+)
+from pants.engine.unions import union
+
+logger = logging.getLogger(__name__)
+
+
+_F = TypeVar("_F", bound=FieldSet)
+
+@union
+@dataclass(frozen=True)
+class PublishPackagesRequest(Generic[_F]):
+    field_set: _F
+    packages: tuple[BuiltPackage, ...]
+
+
+_T = TypeVar("_T", bound=PublishPackagesRequest)
+
+@union
+@dataclass(frozen=True)
+class PublishFieldSet(FieldSet, metaclass=ABCMeta):
+    """The fields necessary to publish an asset from a target.
+
+    Implementing rules should subclass this class ... and return a PublishPackageArtifactRequest as
+    output.
+
+    """
+    publish_request_type: ClassVar[Type[_T]]
+
+    @final
+    def request(self, packages: tuple[BuiltPackage, ...]) -> _T:
+        return self.publish_request_type(field_set=self, packages=packages)
+
+    @final
+    @classmethod
+    def rules(cls) -> tuple[Rule, ...]:
+        return (
+            UnionRule(PublishFieldSet, cls),
+            UnionRule(PublishPackagesRequest, cls.publish_request_type),
+        )
+
+
+@dataclass(frozen=True)
+class PublishPackageProcesses:
+    """Process to run in order to publish named artifact.
+
+    This is returned from implementing rules in response to processing a PublishFieldSet.
+
+    There are multiple names to support processes working on several packages for each process.
+    There are multiple processes to support publishing to several upstream services for each package.
+    """
```

Comment:
Is there a concrete usecase for having one `@rule` produce multiple artifacts like this, as opposed to having multiple `UnionRule`s for the different artifacts it wants to produce? It doesn't seem like a case that needs optimization from a performance perspective, so I wonder whether you had a semantic reason...

Reply:
Yes, no not for optimization, but rather that I map All `BuiltPackage` for a given target to each publish field set for the same. And the publish step may need to invoke multiple processes and the names field is for each artifact in every package being published. In other words, in the simples case, there'll be one process, with a list of names that the process publishes. But it will be up to the rule to decide how to partition the work.

Reply:
Ok... if this is just an implementation detail of the goal, rather than something that a publish backend implementer needs to worry about, that's probably fine. I would emphasize that in the docs though probably, as currently an implementer might think that they needed to grok this.

Reply:
Ok, I'll look over that part. I wrote it in an attempt to make it clearer for myself to begin with :D